### PR TITLE
Diff current file

### PR DESCRIPTION
**Pull request by NicoSantangelo**
Hello!

This PR aims to add a `diff` only for the active file. It's the same view as the current `git: diff` but only showing the current file.
It also adds the same feature inside the status view. I don't know if it's useful but my use case is seeing in a quick glimpse all the changes of a file, before making a commit.

The PR it's not finished because all the code inside `GsStatusDiffCommand` it's duplicated from `GsStatusDiffInlineCommand`. I didn't know how extract it keeping consistency across the project.

Let me know if it's helpful and if you need anything!

### CODE REVIEW

File: core/interfaces/status.py [hunk 1]
```diff
@@ -299,6 +299,63 @@ def load_inline_diff_windows(self, non_cached_files, cached_files):
...
+        non_cached_sections = status_view_section_ranges[self.view.id()][:3]
+        non_cached_lines = util.view.get_lines_from_regions(
+            self.view,
+            self.view.sel(),
+            valid_ranges=non_cached_sections
+        )
+        non_cached_files = (
+            os.path.join(self.repo_path, line.strip())
+            for line in non_cached_lines
+            if line[:4] == "    ")
```
Comment (divmain) [inline core/interfaces/status.py]: Style: place the paren on the next line.

File: core/interfaces/status.py [hunk 2]
```diff
@@ -299,6 +299,63 @@ def load_inline_diff_windows(self, non_cached_files, cached_files):
...
+        cached_sections = status_view_section_ranges[self.view.id()][3:]
+        cached_lines = util.view.get_lines_from_regions(
+            self.view,
+            self.view.sel(),
+            valid_ranges=cached_sections
+        )
+        cached_files = (
+            os.path.join(self.repo_path, line.strip())
+            for line in cached_lines
+            if line[:4] == "    ")
```
Comment (divmain) [inline core/interfaces/status.py]: Ditto.

File: core/commands/diff.py [hunk 3]
```diff
@@ -49,7 +57,8 @@ class GsDiffRefreshCommand(TextCommand, GitCommand):
 
     def run(self, edit, cursors=None):
         in_cached_mode = self.view.settings().get("git_savvy.diff_view.in_cached_mode")
-        stdout = self.git("diff", "--cached" if in_cached_mode else None)
+        file_path = self.view.settings().get("git_savvy.file_path")
+        stdout = self.git("diff", "--cached" if in_cached_mode else None, file_path)
```
Comment (divmain) [inline core/commands/diff.py]: `self.file_path` will evaluate properly if `git_savvy.file_path` is set on the view.
```python
stdout = self.git("diff", "--cached" if in_cached_mode else None, self.file_path)
```

File: Default.sublime-commands [hunk 4]
```diff
@@ -118,6 +118,16 @@
         "args": { "in_cached_mode": true }
     },
     {
+        "caption": "git: diff current file",
+        "command": "gs_diff",
+        "args": { "in_current_file": true }
+    },
+    {
+        "caption": "git: diff cached current file",
```
Comment (divmain) [inline Default.sublime-commands]: Change to `git: diff current file (cached)` to be consistent with other palette commands.

File: core/interfaces/status.py [hunk 5]
```diff
@@ -299,6 +299,63 @@ def load_inline_diff_windows(self, non_cached_files, cached_files):
...
+
+        sublime.set_timeout_async(
+            partial(self.load_diff_windows, non_cached_files, cached_files), 0)
```
Comment (divmain) [inline core/interfaces/status.py]: Prefer a `lambda` here to be consistent with rest of code base.

File: core/commands/diff.py [hunk 6]
```diff
@@ -28,14 +28,22 @@ class GsDiffCommand(WindowCommand, GitCommand):
     def run(self, **kwargs):
         sublime.set_timeout_async(lambda: self.run_async(**kwargs), 0)
 
-    def run_async(self, in_cached_mode=False):
-        repo_path = self.repo_path
+    def run_async(self, settings=None, in_cached_mode=False, in_current_file=False):
```
Comment (divmain) [inline core/commands/diff.py]: Change the method signature to the following:

```python
def run_async(self, in_cached_mode=False, file_path=None, current_file=False)
```

Then below,
```python
if current_file:
    file_path = self.file_path
```

File: core/commands/diff.py [hunk 7]
```diff
@@ -28,14 +28,22 @@ class GsDiffCommand(WindowCommand, GitCommand):
...
+            repo_path = settings["git_savvy.repo_path"]
+
         diff_view = util.view.get_read_only_view(self, "diff")
         title = (DIFF_CACHED_TITLE if in_cached_mode else DIFF_TITLE).format(os.path.basename(repo_path))
         diff_view.set_name(title)
         diff_view.set_syntax_file("Packages/Diff/Diff.tmLanguage")
         diff_view.settings().set("git_savvy.repo_path", repo_path)
         diff_view.settings().set("git_savvy.diff_view.in_cached_mode", in_cached_mode)
+        if in_current_file:
+            diff_view.settings().set("git_savvy.file_path", file_path)
```
Comment (divmain) [inline core/commands/diff.py]: And change this to:
```python
diff_view.settings().set("git_savvy.file_path", file_path)
```
If `file_path` can be `None` and nothing will break.

File: core/commands/diff.py [hunk 8]
```diff
@@ -28,14 +28,22 @@ class GsDiffCommand(WindowCommand, GitCommand):
...
-    def run_async(self, in_cached_mode=False):
-        repo_path = self.repo_path
+    def run_async(self, settings=None, in_cached_mode=False, in_current_file=False):
+        if settings is None:
+            file_path = self.file_path
+            repo_path = self.repo_path
+        else:
+            file_path = settings["git_savvy.file_path"]
+            repo_path = settings["git_savvy.repo_path"]
+
```
Comment (divmain) [inline core/commands/diff.py]: This block of code can go away.

File: core/commands/diff.py [hunk 9]
```diff
@@ -28,14 +28,22 @@ class GsDiffCommand(WindowCommand, GitCommand):
     def run(self, **kwargs):
         sublime.set_timeout_async(lambda: self.run_async(**kwargs), 0)
 
-    def run_async(self, in_cached_mode=False):
-        repo_path = self.repo_path
```
Comment (divmain) [inline core/commands/diff.py]: And this line should stay as-is.

File: core/interfaces/status.py [hunk 10]
```diff
@@ -299,6 +299,63 @@ def load_inline_diff_windows(self, non_cached_files, cached_files):
...
+
+    def load_diff_windows(self, non_cached_files, cached_files):
+        for fpath in non_cached_files:
+            settings = {
+                "git_savvy.file_path": fpath,
+                "git_savvy.repo_path": self.repo_path,
+            }
+            self.view.window().run_command("gs_diff", {
+                "settings": settings,
```
Comment (divmain) [inline core/interfaces/status.py]: You can remove the `settings` param and instead pass `"file_path": fpath`.

**Comment by divmain** [comment]
@NicoSantangelo, had a lot of feedback, but hopefully they are relatively minor changes.  Thanks for the PR, looking forward to merging this!

**Comment by NicoSantangelo** [comment]
Thanks again for taking the time to review my changes. 
I'll let you know when I'm done!

**Comment by NicoSantangelo** [comment]
I didn't change the parens and lambda because I've copied that code from `GsStatusDiffInlineCommand` [here](https://github.com/divmain/GitSavvy/blob/master/core/interfaces/status.py#L227). Do you want me to change both?

My idea was to extract this block of code:

````python
    def run(self, edit):
        # Unstaged, Untracked, and Conflicts
        non_cached_sections = status_view_section_ranges[self.view.id()][:3]
        non_cached_lines = util.view.get_lines_from_regions(
            self.view,
            self.view.sel(),
            valid_ranges=non_cached_sections
        )
        non_cached_files = (
            os.path.join(self.repo_path, line.strip())
            for line in non_cached_lines
            if line[:4] == "    ")

        # Staged
        cached_sections = status_view_section_ranges[self.view.id()][3:]
        cached_lines = util.view.get_lines_from_regions(
            self.view,
            self.view.sel(),
            valid_ranges=cached_sections
        )
        cached_files = (
            os.path.join(self.repo_path, line.strip())
            for line in cached_lines
            if line[:4] == "    ")

        sublime.set_timeout_async(
            partial(self.load_inline_diff_windows, non_cached_files, cached_files), 0)
````

for both `GsStatusDiffInlineCommand` and `GsStatusDiffCommand`. But I don't know where to put the code without being inconsistent.

**Comment by divmain** [comment]
Yeah, go ahead and change it in both places, and remove `from functools import partial`.  I probably wrote that pretty tired - generally, would like to use the same mechanism unless there's a good reason.

As far as making this more DRY, that can be a separate effort.  I'm noticing a handful of places where the repetition occurs, and I'm thinking on how I want to organize the code.

**Comment by NicoSantangelo** [comment]
I finally finished.
If you need me to change anything else let me know!

**Comment by divmain** [comment]
@NicoSantangelo the PR looks strange - that may be due to the ongoing GitHub DDoS.  In any case, your work is merged into master manually, as there was a merge conflict.  Thanks again for this!

**Comment by NicoSantangelo** [comment]
@divmain Wow this is weird.
I hope it wasn't much trouble.

**Comment by divmain** [comment]
Master looks OK and your work is present there, so no worries!

**Pull request merged**